internal/encryption_backend/pkcs11.go

```diff
@@ -0,0 +1,274 @@
+
+import (
+	"crypto/rand"
+	"fmt"
+
+	"github.com/miekg/pkcs11"
+	"go.uber.org/zap"
+)
+
+// PKCS11Provider defines the interface for PKCS11 operations needed by our backend
+type PKCS11Provider interface {
+	Initialize() error
+	Finalize() error
+	GetSlotList(bool) ([]uint, error)
+	OpenSession(uint, uint) (pkcs11.SessionHandle, error)
+	CloseSession(pkcs11.SessionHandle) error
+	Login(pkcs11.SessionHandle, uint, string) error
+	Logout(pkcs11.SessionHandle) error
+	FindObjectsInit(pkcs11.SessionHandle, []*pkcs11.Attribute) error
+	FindObjects(pkcs11.SessionHandle, int) ([]pkcs11.ObjectHandle, bool, error)
+	FindObjectsFinal(pkcs11.SessionHandle) error
+	EncryptInit(pkcs11.SessionHandle, []*pkcs11.Mechanism, pkcs11.ObjectHandle) error
+	Encrypt(pkcs11.SessionHandle, []byte) ([]byte, error)
+	DecryptInit(pkcs11.SessionHandle, []*pkcs11.Mechanism, pkcs11.ObjectHandle) error
+	Decrypt(pkcs11.SessionHandle, []byte) ([]byte, error)
+}
+
+// realPKCS11Provider implements PKCS11Provider using the real pkcs11.Ctx
+type realPKCS11Provider struct {
+	ctx *pkcs11.Ctx
+}
+
+func (p *realPKCS11Provider) Initialize() error { return p.ctx.Initialize() }
+func (p *realPKCS11Provider) Finalize() error   { return p.ctx.Finalize() }
+func (p *realPKCS11Provider) GetSlotList(tokenPresent bool) ([]uint, error) {
+	return p.ctx.GetSlotList(tokenPresent)
+}
+func (p *realPKCS11Provider) OpenSession(slotID uint, flags uint) (pkcs11.SessionHandle, error) {
+	return p.ctx.OpenSession(slotID, flags)
+}
+func (p *realPKCS11Provider) CloseSession(sh pkcs11.SessionHandle) error {
+	return p.ctx.CloseSession(sh)
+}
+func (p *realPKCS11Provider) Login(sh pkcs11.SessionHandle, ut uint, pin string) error {
+	return p.ctx.Login(sh, ut, pin)
+}
+func (p *realPKCS11Provider) Logout(sh pkcs11.SessionHandle) error { return p.ctx.Logout(sh) }
+func (p *realPKCS11Provider) FindObjectsInit(sh pkcs11.SessionHandle, temp []*pkcs11.Attribute) error {
+	return p.ctx.FindObjectsInit(sh, temp)
+}
+func (p *realPKCS11Provider) FindObjects(sh pkcs11.SessionHandle, max int) ([]pkcs11.ObjectHandle, bool, error) {
+	return p.ctx.FindObjects(sh, max)
+}
+func (p *realPKCS11Provider) FindObjectsFinal(sh pkcs11.SessionHandle) error {
+	return p.ctx.FindObjectsFinal(sh)
+}
+func (p *realPKCS11Provider) EncryptInit(sh pkcs11.SessionHandle, m []*pkcs11.Mechanism, o pkcs11.ObjectHandle) error {
+	return p.ctx.EncryptInit(sh, m, o)
+}
+func (p *realPKCS11Provider) Encrypt(sh pkcs11.SessionHandle, message []byte) ([]byte, error) {
+	return p.ctx.Encrypt(sh, message)
+}
+func (p *realPKCS11Provider) DecryptInit(sh pkcs11.SessionHandle, m []*pkcs11.Mechanism, o pkcs11.ObjectHandle) error {
+	return p.ctx.DecryptInit(sh, m, o)
+}
+func (p *realPKCS11Provider) Decrypt(sh pkcs11.SessionHandle, message []byte) ([]byte, error) {
+	return p.ctx.Decrypt(sh, message)
+}
+
+// PKCS11Backend implements EncryptionBackend using the PKCS11 protocol for HSMs.
+type PKCS11Backend struct {
+	ctx    PKCS11Provider
```

Comment:
```suggestion
	pkcs11Provider    PKCS11Provider
```

Generally I'd avoid using `ctx` or `context` as variables, those are overly generic and encourage us to put anything in there. Also, those are generally used for `ctx context.Context` which we'll likely introduce in the future if/when we add tracing support.

Reply:
Renamed